pkg/reconciler/instances/eventing/preaction/handletogglejsfilestorage.go

```diff
@@ -0,0 +1,205 @@
+	if err != nil {
+		return err
+	}
+
+	statefulSet, err := clientSet.AppsV1().StatefulSets(namespace).Get(context.Context, statefulSetName, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.With(log.KeyReason, "Nats StatefulSet not found").Info("Step skipped")
+			return nil
+		}
+		return err
+	}
+
+	tracker, err := progress.NewProgressTracker(clientSet, logger, progress.Config{Interval: progressTrackerInterval, Timeout: progressTrackerTimeout})
+	if err != nil {
+		return err
+	}
+
+	comp := chart.NewComponentBuilder(context.Task.Version, eventingComponentName).
+		WithConfiguration(context.Task.Configuration).
+		WithNamespace(namespace).
+		Build()
+
+	chartValues, err := context.ChartProvider.Configuration(comp)
+	if err != nil {
+		return err
+	}
+
+	data, err := yaml.Marshal(chartValues)
+	if err != nil {
+		return err
+	}
+
+	values := &Config{}
+	if err := yaml.Unmarshal(data, &values); err != nil {
+		return err
+	}
+
+	// fetch Jetstream configuration
+	jetstreamEnabled := fmt.Sprint(values.Global.Jetstream.Enabled) == "true"
+	storageType := fmt.Sprint(values.Global.Jetstream.Storage)
+
+	// check if the Jetstream PVC is in place
+	volumeClaimTemplateFound := false
+	for _, volumeClaimTemplate := range statefulSet.Spec.VolumeClaimTemplates {
+		if strings.EqualFold(volumeClaimTemplate.Name, volumeClaimName) {
+			volumeClaimTemplateFound = true
+		}
+	}
+
+	//Scenario 1: migrate from NATS to NATS Jetstream with file storage
+	//as the statefulSet's spec is immutable, we cannot patch in order to add a PVC, hence statefulSet must be recreated
+	if jetstreamEnabled && strings.EqualFold(fileStorageType, fmt.Sprint(storageType)) {
+		if !volumeClaimTemplateFound {
+			if err := deleteNatsStatefulSet(context, clientSet, tracker, logger, statefulSet, false); err != nil {
+				return err
+			}
+		} else {
+			logger.With(log.KeyReason, "Nats StatefulSet already contains a PVC").Info("Step skipped")
+		}
+		return nil
+	}
+
+	// Scenario 2: migrate from NATS Jetstream with file storage to NATS
+	// as the statefulSet's spec is immutable, we cannot patch it. Hence, in order to remove a PVC, the StatefulSet must be recreated
+	if !jetstreamEnabled && volumeClaimTemplateFound {
+		if err := deleteNatsStatefulSet(context, clientSet, tracker, logger, statefulSet, true); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	logger.With(log.KeyReason, "No actions needed").Info("Step skipped")
+	return nil
+}
+
+// deleteNatsStatefulSet delete the Nats StatefulSet and optionally its assigned PVC.
+func deleteNatsStatefulSet(context *service.ActionContext, clientSet k8s.Interface, tracker *progress.Tracker, logger *zap.SugaredLogger, statefulSet *appsv1.StatefulSet, deletePVC bool) error {
+	// In case of switching from Nats Jestream to NATS the PVCs assigned to the Nats Stateful set should be deleted
+	if deletePVC {
+		err := deleteNatsJsPVC(context, statefulSet, clientSet, logger)
```

Comment:
Can a PVC be deleted if its bind to some StatefulSet?
What if we first delete the StatefulSet and then delete the PVC?
Please test and verify this workflow?